### DISCUSSION TITLE
Plugins when using NixOS home-manager

### DISCUSSION THREAD

**Discussion by jacanchaplais** [sxyazi/yazi#1276] [1. Q&A]
Hi there!

I am fairly new to both NixOS and yazi, so hopefully my n00b status isn't too aggravating. I have yazi installed and configured via home manager

```nix
  programs.yazi = {
    enable = true;
    enableZshIntegration = true;
    settings = {
      manager = {
        show_hidden = true;
      };
      opener.edit = [
        { run = "nvim \"$@\""; block = true; }
      ];
    };
    keymap = {
      manager.prepend_keymap = [
        { run = "close"; on = [ "<C-q>" ]; }
        { run = "yank --cut"; on = [ "d" ]; }
        { run = "remove --force"; on = [ "D" ]; }
        { run = "remove --permanently"; on = [ "X" ]; }
      ];
    };
  };
```

but I see, unlike in configuration.nix, there's no option for plugins. I wanted to add a bookmarks plugin, so I switched over to configuration.nix. I tried to get it to use fetchFromGitHub, and then pass that as an attribute to the plugins option, but it simply wouldn't build (apparently it wasn't happy with me trying to coerce a function to a string? I don't really know what these fetcher functions return tbh, the documentation sent me in circles).

Long story short, my question is this: is there a simple way to use plugins with home-manager? It seems a shame to lose the plugin ecosystem when using nix, but I can't seem to work it out in configuration.nix, and home-manager doesn't have the option at all, right?

**Comment by jacanchaplais** [comment]
For context, this is what I was doing when I was trying without home-manager.

The top of my configuration.nix file:

```nix
{ config, pkgs, inputs, ... }:

let
  yaziBookmarkSrc = pkgs.fetchFromGitHub {
    owner  = "dedukun";
    repo   = "bookmarks.yazi";
    rev    = "08b3f85c5d52656157a55e10410050b042f5b314";
    sha256 = "sha256-t3MKmmNABMMphnfpIOQiSfn34HwNo8RkWr7+jeD7xfI=";
  };
in
{
  imports =
    [ # Include the results of the hardware scan.
      ./hardware-configuration.nix
      inputs.home-manager.nixosModules.default
    ];
```

The yazi config:

```nix
  programs.yazi = {
    enable = true;
    # enableZshIntegration = true;
    settings = {
      keymap = {
        manager.prepend_keymap = [
          { run = "close"; on = [ "<C-q>" ]; }
          { run = "yank --cut"; on = [ "d" ]; }
          { run = "remove --force"; on = [ "D" ]; }
          { run = "remove --permanently"; on = [ "X" ]; }
          {
            run = "plugin bookmarks --args=save";
            on = [ "m" ];
            desc = "Save current position as a bookmark";
          }
          {
            run = "plugin bookmarks --args=jump";
            on = [ "'" ];
            desc = "Jump to a bookmark";
          }
          {
            run = "plugin bookmarks --args=delete";
            on = [ "b" "d" ];
            desc = "Delete a bookmark";
          }
          {
            run = "plugin bookmarks --args=delete_all";
            on = [ "b" "D" ];
            desc = "Delete all bookmarks";
          }
        ];
      };
      yazi = {
        manager = {
          show_hidden = true;
        };
        opener.edit = [
          { run = "nvim \"$@\""; block = true; }
        ];
      };
    };
    plugins = {
      bookmarks = yaziBookmarkSrc;
    };
  };
```

Just in case it helps :)

**Comment by sxyazi** [comment]
Hi, I've added an example to the docs here https://github.com/yazi-rs/yazi-rs.github.io/commit/5388f1230249c12f95e495d0a38dc69decc4305b to demonstrate configuring Yazi with home-manager, hope it helps

**Reply by jacanchaplais** [reply]
Many thanks for the fast reply! Unfortunately, the programs.yazi in home-manager doesn't appear to have a plugins option? I can't see one in the man page, and running your code example gives:

```
       error: The option `home-manager.users.jacan.programs.yazi.plugins' does not exist. Definition values:
       - In `/nix/store/j43ma9b1fiqhncircgckqglsbdcgsvdq-source/.config/nixos/hosts/default/configuration.nix':
```

**Reply by jacanchaplais** [reply]
(it does have this option in the standard configuration.nix, without home-manager)

**Reply by VictorVSa** [reply]
I can also see a mix set of configuration options between home-manager and non-manager, some options are available for one but not the other, like plugins (only non-hm) and enableZshIntegration (only hm)

HM: https://mynixos.com/home-manager/options/programs.yazi

Non-HM: https://search.nixos.org/options?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=yazi

**Reply by VictorVSa** [reply]
Temporary hack I've written:

```nix
  xdg = {
    enable = true;
    configFile = {
        "yazi/init.lua".text = ''
          require("full-border"):setup()
        '';
      }
      // (lib.attrsets.mapAttrs'
        (name: value: lib.attrsets.nameValuePair "yazi/plugins/${name}.yazi" { source = value; })
        {
          chmod = "${plugins-repo}/chmod.yazi";
          full-border = "${plugins-repo}/full-border.yazi";
          max-preview = "${plugins-repo}/max-preview.yazi";
          diff = "${plugins-repo}/diff.yazi";
        }
      );
  };

```

**Reply by MangoCubes** [reply]
I fixed this! The unstable channel has those options, but 24.05 does not.
See this (24.05):
https://home-manager-options.extranix.com/?query=yazi&release=release-24.05
and this (unstable):
https://home-manager-options.extranix.com/?query=yazi&release=master

**Comment by thom-cameron** [comment]
Hey, has anyone figured out how to write plugin code directly into your config file? For example, to add the code for file navigation wraparound as shown on the [tips page](https://yazi-rs.github.io/docs/tips#navigation-wraparound).

I've tried writeScriptDir and other similar builders but can't get anything working.
